### PR TITLE
Miscellaneous changes for HM type inference and QEP 2

### PR DESCRIPTION
**Pull request by samarsha**
* Add parsing for empty array literals.
* Support sized array literals in attribute arguments.
* Replace warning when a type parameter doesn't occur in the argument with a warning when the type parameter is unused (in both the argument and the return types).
* Add tests for integral, iterable, numeric, semigroup, and unwrap constraints.

### CODE REVIEW

File: src/QsCompiler/Tests.Compiler/SyntaxTests.fs [hunk 1]
```diff
@@ -403,8 +403,8 @@ let ``Identifier tests`` () =
 [<Fact>]
 let ``Complex literal tests`` () =
     [
-        "[]", true, toExpr InvalidExpr, [ Error ErrorCode.EmptyValueArray ]
-        "[ ]", true, toExpr InvalidExpr, [ Error ErrorCode.EmptyValueArray ]
+        "[]", true, toArray [], []
```
Comment (bettinaheim) [inline src/QsCompiler/Tests.Compiler/SyntaxTests.fs:406]: Are there some tests somewhere for the sized arrays?

**Review by bettinaheim — CHANGES_REQUESTED** [review]
- I am looking for an override of OnSizedArray in the QIR emission and don't see it yet. What's the plan for that?
- Also, in CompilerDataStructures.cs, I see a couple of QsExpressionKindNewArray without a matching QsExpressionKindSizedArray - is there something missing there? In BondSchemaTranslator.cs, e.g. in QsExpressionKindComposition NewArray is listed but not SizedArray. I would also expect "SizedArray" to occur in CompilerObjectTranslator. @cesarzc Could you take a look at the bond schemas to have a second set of eyes on that?
- In ConstantPropagation.fs in shouldPropagate there is NewArray listed but not SizedArray, same in Evaluation.fs, same in HelperFunctions.fs in isLiteral
- In NamespaceManager.fs in ExpressionHash I would also expect to find SizedArray
- There is also an overload missing in QsharpCodeOutput.cs

Also, is there a matching PR on the runtime repo to add support in the C# generation?

**Comment by samarsha** [comment]
> * I am looking for an override of OnSizedArray in the QIR emission and don't see it yet. What's the plan for that?
> * Also, in CompilerDataStructures.cs, I see a couple of QsExpressionKindNewArray without a matching QsExpressionKindSizedArray - is there something missing there? In BondSchemaTranslator.cs, e.g. in QsExpressionKindComposition NewArray is listed but not SizedArray. I would also expect "SizedArray" to occur in CompilerObjectTranslator. @cesarzc Could you take a look at the bond schemas to have a second set of eyes on that?
> * In ConstantPropagation.fs in shouldPropagate there is NewArray listed but not SizedArray, same in Evaluation.fs, same in HelperFunctions.fs in isLiteral
> * In NamespaceManager.fs in ExpressionHash I would also expect to find SizedArray
> * There is also an overload missing in QsharpCodeOutput.cs
> 
> Also, is there a matching PR on the runtime repo to add support in the C# generation?

I don't think any of these issues are new in this PR. The `SizedArray` expression kind was added in #856, with corresponding C# generation added in microsoft/qsharp-runtime#507.

**Comment by bettinaheim** [comment]
> > * I am looking for an override of OnSizedArray in the QIR emission and don't see it yet. What's the plan for that?
> > * Also, in CompilerDataStructures.cs, I see a couple of QsExpressionKindNewArray without a matching QsExpressionKindSizedArray - is there something missing there? In BondSchemaTranslator.cs, e.g. in QsExpressionKindComposition NewArray is listed but not SizedArray. I would also expect "SizedArray" to occur in CompilerObjectTranslator. @cesarzc Could you take a look at the bond schemas to have a second set of eyes on that?
> > * In ConstantPropagation.fs in shouldPropagate there is NewArray listed but not SizedArray, same in Evaluation.fs, same in HelperFunctions.fs in isLiteral
> > * In NamespaceManager.fs in ExpressionHash I would also expect to find SizedArray
> > * There is also an overload missing in QsharpCodeOutput.cs
> > 
> > Also, is there a matching PR on the runtime repo to add support in the C# generation?
> 
> I don't think any of these issues are new in this PR. The `SizedArray` expression kind was added in #856, with corresponding C# generation added in [microsoft/qsharp-runtime#507](https://github.com/microsoft/qsharp-runtime/pull/507).

See the edit above (you have beaten me to it with your reply):
Edit: I reviewed (i.e. set changes requested) in the same way as I would for a PR to main - since this is going into a feature branch, I am happy to remove the blocker to split it into several PRs if you prefer. If you want to do that, then I suggest to make a task list in a GitHub issue to make sure the remaining work is tracked and not forgotten.

I am listing the things that are missing for the feature to be complete, since this PR is titled "Final touches for ... QEP 2"; up to you when to address them as long as they are done before the merge to main. Could you please track the items I listed somewhere, if they are not addressed with this PR?

**Comment by samarsha** [comment]
Sorry, the name is misleading. I didn't intend for this to be the final PR.

**Comment by samarsha** [comment]
> Also, in CompilerDataStructures.cs, I see a couple of QsExpressionKindNewArray without a matching QsExpressionKindSizedArray - is there something missing there?

That's intentional. `SizedArray` uses `QsExpressionKindExpressionDouble`.

> In BondSchemaTranslator.cs, e.g. in QsExpressionKindComposition NewArray is listed but not SizedArray. I would also expect "SizedArray" to occur in CompilerObjectTranslator.

Also intentional, due to piggy-backing on the existing `QsExpressionKindExpressionDouble`.

> @cesarzc Could you take a look at the bond schemas to have a second set of eyes on that?

@cesarzc already reviewed the changes to the Bond schema in #856.

**Comment by samarsha** [comment]
> I am looking for an override of OnSizedArray in the QIR emission and don't see it yet. What's the plan for that?

Keep in mind that `SizedArray` was added before QIR generation was even in main. :) It should be added before merging the feature branch, though, yes.

**Review by bettinaheim — APPROVED** [review]

File: src/QsCompiler/Tests.Compiler/SyntaxTests.fs (hunk 1 above)
Comment (samarsha) [reply src/QsCompiler/Tests.Compiler/SyntaxTests.fs:406]: They are further down in the same `Complex literal tests` function.

**Pull request merged**